File: Haskell/data_structures/04-Heapify.hs
import Control.Monad (replicateM)
import Control.Monad.ST
import Control.Monad.Primitive (PrimMonad, PrimState)
import Data.Vector (Vector(..), fromList, thaw)
import qualified Data.Vector.Mutable as MV


convertToIntList :: String -> [Int]
convertToIntList = map read . words


getIndexOfParentChildrenMinimum :: PrimMonad m => MV.MVector (PrimState m) Int -> Int -> Int -> m Int
getIndexOfParentChildrenMinimum arr parentIndex size = do
    let leftChildIx = 2 * parentIndex + 1
    let rightChildIx = leftChildIx + 1
    minIndex <- if leftChildIx < size then do
                    v1 <- MV.read arr leftChildIx
                    v2 <- MV.read arr parentIndex
                    return $ if v1 < v2 then leftChildIx else parentIndex
                else return parentIndex
    if rightChildIx < size then do
        v1 <- MV.read arr rightChildIx
        v2 <- MV.read arr minIndex
        return $ if v1 < v2 then rightChildIx else minIndex
    else return minIndex


siftDown :: PrimMonad m => MV.MVector (PrimState m) Int -> Int -> Int -> m [(Int, Int)]
siftDown arr parentIndex n = do
    let go currentParentIx currentMinIx swaps
            | currentMinIx == currentParentIx = return swaps
            | otherwise = do
                MV.swap arr currentMinIx currentParentIx
                let nextParentIx = currentMinIx
                nextMinIx <- getIndexOfParentChildrenMinimum arr nextParentIx n
                go nextParentIx nextMinIx ((currentParentIx, currentMinIx) : swaps)
    minIndex <- getIndexOfParentChildrenMinimum arr parentIndex n
    go parentIndex minIndex []


createHeap :: PrimMonad m => MV.MVector (PrimState m) Int -> Int -> [(Int, Int)] -> Int -> m [(Int, Int)]
createHeap arr n swaps parentIndex
    | parentIndex < 0 = return $ reverse swaps
    | otherwise = do
        currentSwaps <- siftDown arr parentIndex n
        createHeap arr n (currentSwaps ++ swaps) (parentIndex - 1)


heapify :: Vector Int -> Int -> [(Int, Int)]
heapify vec n = runST $ do
    arr <- thaw vec
    createHeap arr n [] (n `div` 2 - 1)


main :: IO()
main = do
    n <- readLn
    vec <- fmap (fromList . convertToIntList) getLine
    let swaps = heapify vec n
    print $ length swaps
    mapM_ putStrLn (map (\(i, j) -> show i ++ " " ++ show j) swaps)

File: Rust/data_structures/src/bin/04_heapify.rs
use std::io;

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line.split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    line
}

fn get_index_of_parent_children_minimum(array: &[i32], parent_index: usize, size: usize) -> usize {
    let mut min_index = parent_index;
    let left_child_ix = 2*parent_index + 1;
    if left_child_ix < size && array[left_child_ix] < array[min_index] {
        min_index = left_child_ix;
    }
    let right_child_ix = left_child_ix + 1;
    if right_child_ix < size && array[right_child_ix] < array[min_index] {
        min_index = right_child_ix;
    }
    min_index
}

fn sift_down(array: &mut [i32], mut parent_index: usize, size: usize, swaps: &mut Vec<(usize, usize)>) {
    let mut min_index = get_index_of_parent_children_minimum(array, parent_index, size);
    while min_index != parent_index {
        array.swap(min_index, parent_index);
        swaps.push((parent_index, min_index));
        parent_index = min_index;
        min_index = get_index_of_parent_children_minimum(array, parent_index, size);
    }
}

fn heapify(array: &mut [i32], n: usize) -> Vec<(usize, usize)> {
    let mut swaps = Vec::new();
    let mut parent_ix = n / 2;
    loop {
        if parent_ix <= 0 {
            return swaps
        }
        parent_ix -= 1;
        sift_down(array, parent_ix, n, &mut swaps);
    }
}

fn main() {
    let n = read_line().trim().parse().unwrap();
    let mut array = convert_to_int_vector(&read_line());
    let result = heapify(&mut array, n);
    println!("{}", result.len());
    result.iter().for_each(|(i, j)| println!("{i} {j}"));
}
